x-pack/plugins/fleet/server/services/epm/kibana/assets/tag_assets.ts

```diff
@@ -5,24 +5,69 @@
  * 2.0.
  */
 
+import { v5 as uuidv5 } from 'uuid';
+import { uniqBy } from 'lodash';
 import type { SavedObjectsImportSuccess } from '@kbn/core-saved-objects-common';
 import { taggableTypes } from '@kbn/saved-objects-tagging-plugin/common/constants';
 import type { IAssignmentService, ITagsClient } from '@kbn/saved-objects-tagging-plugin/server';
 
 import type { KibanaAssetType } from '../../../../../common';
+import type { PackageSpecTags } from '../../../../types';
+
 import { appContextService } from '../../../app_context';
 
 import type { ArchiveAsset } from './install';
 import { KibanaSavedObjectTypeMapping } from './install';
 
-const TAG_COLOR = '#FFFFFF';
+interface ObjectReference {
+  type: string;
+  id: string;
+}
+interface PackageSpecTagsAssets {
+  tagId: string;
+  assets: ObjectReference[];
+}
+
+interface GroupedAssets {
+  [assetId: string]: { type: string; tags: string[] };
+}
+
+const MANAGED_TAG_COLOR = '#0077CC';
+const PACKAGE_TAG_COLOR = '#4DD2CA';
 const MANAGED_TAG_NAME = 'Managed';
 const LEGACY_MANAGED_TAG_ID = 'managed';
 
+// the tag service only accepts 6-digits hex colors
+const TAG_COLORS = [
+  '#FEC514',
+  '#F583B7',
+  '#F04E98',
+  '#00BFB3',
+  '#FEC514',
+  '#BADA55',
+  '#FFA500',
+  '#9696F1',
+  '#D36086',
+  '#54B399',
+  '#AAA8A5',
+  '#A0A0A0',
+];
+
 const getManagedTagId = (spaceId: string) => `fleet-managed-${spaceId}`;
 const getPackageTagId = (spaceId: string, pkgName: string) => `fleet-pkg-${pkgName}-${spaceId}`;
 const getLegacyPackageTagId = (pkgName: string) => pkgName;
 
+export const getPackageSpecTagId = (spaceId: string, pkgName: string, tagName: string) => {
+  // UUID v5 needs a namespace (uuid.DNS) to generate a predictable uuid
+  const uniqueId = uuidv5(`${tagName.toLowerCase()}`, uuidv5.DNS);
+  return `fleet-shared-tag-${pkgName}-${uniqueId}-${spaceId}`;
```

Comment:
The decision of using the uuid came in [this comment](https://github.com/elastic/kibana/issues/152814#issuecomment-1516053663) and [this comment](https://github.com/elastic/kibana/issues/152814#issuecomment-1545856989) to avoid name collisions. 

I can also add the lowercased name, but to be honest this id is only meant to be used in the service and the id is unique (and can be checked with online services that generate uuids), adding the lowercase name would make it very long.

Reply:
Okay, I see your point about the tag name being potentially too long. Collisions shouldn't happen either way as we shouldn't create a new tag if the lowercase name is the same.
I'm fine with keeping it a uuid, it would be nice to add some comments to the unit tests where we are checking the created tag id, to know which id corresponds to which tag name.

Reply:
I modified the tests so that the ids are assigned to variables, this should make the tests more readable.